scripts/linkchecker.py

```diff
@@ -0,0 +1,423 @@
+        norm_path = os.path.normpath(path)
+        if ARGS.no_color:
+            print("File: " + norm_path)
+        else:
+            print(C_CYAN + "File: " + norm_path + C_END)
+        for p in path_output:
+            print(" "*4 + p)
+    return
+
+
+def strip_comments(content):
+    """Manual striping of comments from file content.
+
+    Many localized content pages contain original English content in comments.
+    These comments have to be stripped out before analyzing the links.
+    Doing this using regular expression is difficult. Even the grep tool is
+    not suitable for this use case.
+
+    NOTE: We strived to preserve line numbers when producing the resulted
+    text. This can be useful in future if we want to print out the line
+    numbers for bad links.
+    """
+    result = []
+    in_comment = False
+    for line in content:
+        idx1 = line.find("<!--")
+        idx2 = line.find("-->")
+        if not in_comment:
+            # only care if new comment started
+            if idx1 < 0:
+                result.append(line)
+                continue
+
+            # single line comment
+            if idx2 > 0:
+                result.append(line[:idx1] + line[idx2+4:])
+                continue
+            result.append(line[:idx1])
+            in_comment = True
+            continue
+
+        # already in comment block
+        if idx2 < 0:  # ignore whole line
+            result.append("")
+            continue
+        result.append(line[idx2+4:])
+        in_comment = False
+
+    return result
+
+
+def normalize_filename(name, ftype="markdown"):
+    """Guess the filename based on a link target.
+
+    This function only deals with regular files.
+    """
+    if name.endswith("/"):
+        name = name[:-1]
+    if ftype == "markdown":
+        name += ".md"
+    else:
+        name += ".html"
+    return name
+
+
+def check_file_exists(base, path, ftype="markdown"):
+    """Check if the target file exists.
+
+    NOTE: We build a normalized path using 'base' and 'path' values. Suppose
+    the resulted path string is 'foo/bar', we check if 'foo/bar.md' exists,
+    AND we check if 'foo/bar/_index.md' exists.
+
+    :param base: The base directory to begin with
+    :param path: The link target which is a relative path string
+    :returns: A boolean indicating whether the target file exists.
+    """
+    # NOTE: anchor is ignored, can be a todo item
+    parts = path.split("#")
+
+    fn = normalize_filename(parts[0], ftype=ftype)
+    target = base + fn
+
+    if os.path.isfile(target):
+        return True
+
+    dir_name = base + parts[0]
+    if os.path.isdir(dir_name):
+        if os.path.isfile(dir_name + "/_index.md"):
+            return True
+        if os.path.isfile(dir_name + "/_index.html"):
+            return True
+        # /docs/contribute/style/hugo-shortcodes/ has this
+        if os.path.isfile(dir_name + "/index.md"):
+            return True
```

Comment:
Should these use `os.path.join()`?

Reply:
tried ... but it doesn't always work.

Reply:
If this doesn't work on Windows, maybe we should call that out.

Reply:
I don't have a Windows box at hand for testing. So you are probably correct ... this script needs some tuning to run on Windows.